app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -550,22 +556,103 @@ protected void deliverError(CommCareSetupActivity receiver,
     public boolean onCreateOptionsMenu(Menu menu) {
         super.onCreateOptionsMenu(menu);
         menu.add(0, MODE_ARCHIVE, 0, Localization.get("menu.archive")).setIcon(android.R.drawable.ic_menu_upload);
+        menu.add(0, MODE_SMS, 1, Localization.get("menu.sms")).setIcon(android.R.drawable.ic_menu_upload);
         return true;
     }
 
     @Override
     public boolean onPrepareOptionsMenu(Menu menu) {
         super.onPrepareOptionsMenu(menu);
-        
+
         return true;
     }
-    
+
+    /**
+     * Scan SMS messages for texts with profile references.
+     * @param installAutomatically install automatically if reference is found
+     */
+    public void performSMSInstall(boolean installAutomatically){
+        try {
+            String profileLink = this.scanSMSLinks();
+            if (profileLink != null) {
+                // we found a valid profile link, either start install automatically
+                // or move to READY_TO_INSTALL state
+                Log.v("install", "Performing SMS install with link : " + incomingRef);
+                incomingRef = profileLink;
+                if (installAutomatically) {
+                    startResourceInstall();
+                } else {
+                    uiState = UiState.READY_TO_INSTALL;
+                    Toast.makeText(this, Localization.get("menu.sms.ready"), Toast.LENGTH_LONG).show();
+                }
+            } else {
+                // only notify if this was manually triggered, since most people won't use this
+                if(installAutomatically) {
+                    Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+                }
+            }
+        } catch(SignatureException e){
+            // possibly we want to do something more severe here? could be malicious
+            e.printStackTrace();
+            Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+        }

```

Comment:
Try to localize your try/catch code to wrap the minimal amount of code that needed. In this case, the only statement that throws `SignatureException` is `scanSMSLinks`, so the try body should only include that statement.

This is important because it will help the reader more easily understand what could fail and when.